### DISCUSSION TITLE
What is the alternative for LOGGER.entering() and LOGGER.exiting() from Java logger to log4j2 using slf4j

### DISCUSSION THREAD

**Discussion by mehulparmariitr** [apache/logging-log4j2#2083] [Q&A] [answered]
I am migrating code base from simple java logger to log4j2 using sl4j because of having issue while writing asynchronously to a single log file in multiple threads. I noticed entry and exit methods are not available in sl4j interface. what is the best to achieve entry and exit to print class name, method name and parameters which can be object.

**Comment by ppkarwasz** [comment, marked as answer]
The equivalents of JUL's [`Logger#entering(String, String, Object[])`](https://docs.oracle.com/en/java/javase/17/docs/api/java.logging/java/util/logging/Logger.html#entering(java.lang.String,java.lang.String,java.lang.Object%5B%5D)) and [`Logger#exiting(String, String, Object)`](https://docs.oracle.com/en/java/javase/17/docs/api/java.logging/java/util/logging/Logger.html#exiting(java.lang.String,java.lang.String,java.lang.Object)) are:

 - [`Logger#traceEntry`](https://logging.apache.org/log4j/2.x/javadoc/log4j-api/org/apache/logging/log4j/Logger.html#traceEntry(java.lang.String,java.lang.Object...)),
 - [`Logger#traceExit`](https://logging.apache.org/log4j/2.x/javadoc/log4j-api/org/apache/logging/log4j/Logger.html#traceExit(R)).

Basically these allow you to replace:

```java
private static final java.util.logging.Logger logger = ...

public String computeFoo(String param1, int param2) {
    logger.entering("org.example.MyClass", "computeFoo", new Object[] {param1, param2});
    // do something

    //
   logger.exiting("org.example.MyClass", "computeFoo", result);
   return result;
````

with:

```java
private static final org.apache.logging.log4j.Logger logger = ...

public String computeFoo(String param1, int param2) {
    logger.traceEntry(null, new Object[] {param1, param2});
    // or just
    logger.traceEntry(null, param1, param2);
    // do something

    //
    return logger.traceExit(result);
````

As you can see the location information is **removed**, because most modern logging backends can compute it on the fly. E.g. with Log4j Core (the reference implementation of Log4j API) you can use:

```xml
<PatternLayout pattern="...%C#%M %m%n"/>
```

Normally runtime location information is expensive, but you can pre-compute it as compiletime using the [Log4j Transform Plugin](https://logging.apache.org/log4j/transform/latest/#log4j-transform-maven-plugin).

**Remark**: This solution requires using **Log4j API** instead of SLF4J. The default SLF4J `Logger` has no `entry/exit` methods. If you **really** want to use SLF4J there is a [`slf4j-ext`](https://mvnrepository.com/artifact/org.slf4j/slf4j-ext) artifact with an `XLogger` class. If I am not mistaken the `entry/exit` methods were introduced in [`XLogger`](https://www.javadoc.io/doc/org.slf4j/slf4j-ext/latest/org/slf4j/ext/XLogger.html) class by @rgoers, before he decided to start Log4j 2.x.

**Reply by ppkarwasz** [reply]
Personally I am not a big fan of the `traceEntry/traceExit` methods.

I believe that this functionality should be provided by aspect-oriented programming. We might for example enhance the aforementioned Log4j Transform Plugin to add entry/exit log statements to a configurable set of methods. Would you mind adding a feature request to the [`logging-log4j-transform` repository](https://github.com/apache/logging-log4j-transform/issues)? Please explain there what kind of methods you want to decorate with entry/exit log statements.

**Reply by mehulparmariitr** [reply]
I am building a customLogger on top of log4j2. How will this custom logger know the name of method automatically to pass to [Logger#traceEntry](https://logging.apache.org/log4j/2.x/javadoc/log4j-api/org/apache/logging/log4j/Logger.html#traceEntry(java.lang.String,java.lang.Object...))?

**Reply by ppkarwasz** [reply]
It won't if you use `Logger#traceEntry`.

Classes that wrap a `Logger` instance should only use the methods in [`Logger`](https://logging.apache.org/log4j/2.x/javadoc/log4j-api/org/apache/logging/log4j/Logger.html) and [`ExtendedLogger`](https://logging.apache.org/log4j/2.x/javadoc/log4j-api/org/apache/logging/log4j/spi/ExtendedLogger.html) that have a parameter called `fqcn`. This parameter must contain the fully qualified class name of your logger class and allows Log4j API implementations to find the location information of the caller (module name, class name, method name, source file name, line number, etc.)

The `ExtendedLogger` interface is specifically provided for integrators and it is hidden from normal users. However every `Logger` implements `ExtendedLogger` (you can just cast it).

**BTW**: I just opened an issue on OpenRewrite (openrewrite/rewrite-logging-frameworks#132) to add a JUL to Log4j API conversion recipe. IMHO this might be a faster and more stable solution to your problem.

**Comment by ppkarwasz** [comment]
Check my OpenRewrite PR for a basic JUL to Log4j API conversion: openrewrite/rewrite-logging-frameworks#133

**Reply by ppkarwasz** [reply]
@mehulparmariitr,

Tim already accepted my `org.openrewrite.java.logging.log4j.JulToLog4j` recipe. Can you check the [OpenRewrite snapshot version](https://docs.openrewrite.org/reference/snapshot-instructions) to see if it helps in your migration task?
If you prefer a stable release, `rewrite-logging-frameworks` have a release every 2 weeks.
